extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/exec/SegmentLoadWaiter.java

```diff
@@ -0,0 +1,304 @@
+import java.util.TreeSet;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Class that periodically checks with the broker if all the segments generated are loaded by querying the sys table
+ * and blocks till it is complete. This will account for and not wait for segments that would never be loaded due to
+ * load rules. Should only be called if the query generates new segments or tombstones.
+ * <br>
+ * If an exception is thrown during operation, this will simply log the exception and exit without failing the task,
+ * since the segments have already been published successfully, and should be loaded eventually.
+ * <br>
+ * If the segments are not loaded within {@link #TIMEOUT_DURATION_MILLIS} milliseconds, this logs a warning and exits
+ * for the same reason.
+ */
+public class SegmentLoadWaiter
+{
+  private static final Logger log = new Logger(SegmentLoadWaiter.class);
+  private static final long INITIAL_SLEEP_DURATION_MILLIS = TimeUnit.SECONDS.toMillis(5);
+  private static final long SLEEP_DURATION_MILLIS = TimeUnit.SECONDS.toMillis(5);
+  private static final long TIMEOUT_DURATION_MILLIS = TimeUnit.MINUTES.toMillis(10);
+  private static final String LOAD_QUERY = "SELECT COUNT(*) AS totalSegments,\n"
+                                           + "COUNT(*) FILTER (WHERE is_available = 0 AND is_published = 1 AND replication_factor != 0) AS loadingSegments\n"
+                                           + "FROM sys.segments\n"
+                                           + "WHERE datasource = '%s' AND is_overshadowed = 0 AND version = '%s'";
+
+  private final BrokerClient brokerClient;
+  private final ObjectMapper objectMapper;
+  // Map of version vs latest load status.
+  private final Map<String, VersionLoadStatus> versionToLoadStatusMap;
+  private final String datasource;
+  private final Set<String> versionsToAwait;
+  private final boolean doWait;
+  private volatile SegmentLoadWaiterStatus status;
+
+  public SegmentLoadWaiter(BrokerClient brokerClient, ObjectMapper objectMapper, String datasource, Set<String> versionsToAwait, int initialSegmentCount, boolean doWait)
+  {
+    this.brokerClient = brokerClient;
+    this.objectMapper = objectMapper;
+    this.datasource = datasource;
+    this.versionsToAwait = new TreeSet<>(versionsToAwait);
+    this.versionToLoadStatusMap = new HashMap<>();
+    this.status = new SegmentLoadWaiterStatus(State.INIT, null, 0, initialSegmentCount, initialSegmentCount);
+    this.doWait = doWait;
+  }
+
+  /**
+   * Uses broker client to check if all segments created by the ingestion have been loaded and updates the {@link #status)}
+   * periodically.
+   * <br>
+   * If an exception is thrown during operation, this will log the exception and return without failing the task,
+   * since the segments have already been published successfully, and should be loaded eventually.
+   * <br>
+   * Only expected to be called from the main controller thread.
+   */
+  public void waitForSegmentsToLoad()
```

Comment:
We would want the experience on the console to be realtime ie like how counters currently work so that the console can render the waiting segment status progress to the end user. 

With this approach the main thread looks to be blocked until the segment loading is complete and the info is only included in the task report once the call returns?
You could add a method which check if the segment loading is complete. If not get the status of loading and write it in the task report. 
We would want to do this until the segment loading is completed.

Reply:
Discussed offline, the controller thread updates the status periodically without blocking. This status is included from the liveReports() in the controller. I have tested this out as well and the realtime information is available to the console